### PR TITLE
Track Metrics About Tokenization Results

### PR DESCRIPTION
**Pull request by snova-zoltanc**
## Summary
There are a lot of metadata details about the dataset that would be useful to know, but it would require writing non trivial python script to extract that from the HDF5 files. So lets keep track of all this metadata and print it so the use can learn more about their resulting dataset.

How the output looks like
image

Things to note: 
1. I will add a follow up PR to log this in a log file, so this information is not lost in the terminal. 
2. There was a major bug in the code during greedy packing, that mean't that if two sequences in a row did not fit, it would add a sequence of all PADDING tokens. This PR fixes this, and the associated test cases. The issue probably DID NOT impact previous tokenization runs because by default later steps drops sequences without completions, so these sequences are dropped unless a flag was passed in to skip this all prompt sequence dropping stage.
3. I updated the README with a table that describes the metrics with the definition and intuition behind them. 

## PR Checklist
- [ X] My PR is less than 500 lines of code
- [X ] I have added sufficient comment as docstrings in my code
- [X ] I have made corresponding changes to the documentation

- [ ] I have written unit-tests to test all of my code

### CODE REVIEW

File: generative_data_prep/processors/metrics.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+            ["Sequence Utilization", self._to_str_percent(self.sequence_utilization)],
+            ["Sequence Completion Utilization", self._to_str_percent(self.sequence_completion_utilization)]
+        ]
+        ret = "=================================================="
+        for name, value in table:
+            ret += "\n" + f"{name}: {value}"
+        ret += "\n" + "=================================================="
+        return ret
+
+        # return tabulate(table, "fancy_outline")
```
Comment (snova-zoltanc) [inline generative_data_prep/processors/metrics.py:129]: I want to replace this with tabulate to get a nice table print, but I am not sure how to get the dependency working. I asked @snova-ranjanl to help

**Comment by snova-zoltanc** [comment]
image

Create a nice table for tracking dataset metrics.

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -33,6 +33,7 @@ repos:
     hooks:
     - id: mypy
       exclude: ^(tests\/.*|docs\/.*)$
+      additional_dependencies: [types-tabulate]
```
Comment (snova-ranjanl) [inline .pre-commit-config.yaml:36]: we should be able to add this setting to `pyproject.toml` instead.

File: generative_data_prep/data_prep/data_prep.py [hunk 3]
```diff
@@ -93,3 +93,6 @@ def data_prep_main(
             for line in reader:
                 hdf5_text_buffer.write(article_tokenizer(line))
             hdf5_text_buffer.write(article_tokenizer(None))
+
+    # lets return an object here that stores all the metrics we want to track
```
Comment (snova-ranjanl) [inline generative_data_prep/data_prep/data_prep.py:97]: nit: should this comment go to the function-docstrings instead?

**Review by snova-ranjanl — CHANGES_REQUESTED** [review]
Looks good overall, and needs a few tweaks.

Regarding the `hdf5` binary files, do we need these here or are these merely artifacts?

**Comment by snova-ranjanl** [comment]
@snova-zoltanc actually, your `additional_dependency` for `tabulate` is fine to be in the `pre-commit` config file.

**Review by snova-ranjanl — APPROVED** [review]
looks good if it passes all CI

File: generative_data_prep/tokenized_line/tokenized_line.py [hunk 4]
```diff
@@ -213,6 +213,21 @@ def free_tokens(self):
...
+        return sum(token.token_type_id == TokenTypeIds.PROMPT for token in self.tokens)
+
+    @property
+    def completion_tokens(self):
+        """The number of completion tokens in this sequence."""
+        return sum(token.token_type_id in [TokenTypeIds.COMPLETION, TokenTypeIds.SEP] for token in self.tokens)
+
+    @property
+    def pad_tokens(self):
+        """The number of completion tokens in this sequence."""
```
Comment (snova-virens) [inline generative_data_prep/tokenized_line/tokenized_line.py:228]: typo in docstring

File: generative_data_prep/processors/metrics.py [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+        self.articles: int = 0  # how many articles are in the input dataset
+        self.tokens_dropped_from_packing: int = 0  # how many tokens are dropped because of packing
+        self.tokens_dropped_from_all_prompt: int = (
+            0  # how many tokens are dropped because no completions in the entire sequence
+        )
+
+    def __iadd__(self: MetricsSubClass, new_metrics: "Metrics") -> MetricsSubClass:
+        """Implement += for Metrics."""
+        self.input_tokens += new_metrics.input_tokens
+        self.total_tokens += new_metrics.total_tokens
```
Comment (snova-virens) [inline generative_data_prep/processors/metrics.py:51]: could we rename this to self.output_tokens?

File: generative_data_prep/processors/article_tokenizer.py [hunk 6]
```diff
@@ -182,8 +205,11 @@ def _remove_prompt_only_sequences(self, tokenized_lines: List[TokenizedSequence]
                     )
                     self.logged_prompt_only_warn_msg_postpack = True
                 continue
-            filtered_lines.append(line)
-        return filtered_lines
+            filtered_sequences.append(line)
```
Comment (snova-virens) [inline generative_data_prep/processors/article_tokenizer.py:208]: could we add a brief comment explaining the bottom two lines

File: generative_data_prep/processors/article_tokenizer.py [hunk 7]
```diff
@@ -159,19 +179,22 @@ def __call__(self, article: Optional[str]) -> List[TokenizedSequence]:
...
-            tokenized_lines: List of TokenizedLines
+            tokenized_sequences: List of TokenizedSequence
         Returns:
             Original list with prompt-only sequences filtered out
         """
-        filtered_lines = []
-        for line in tokenized_lines:
+        filtered_sequences = []
+        for line in tokenized_sequences:
             if TokenTypeIds.COMPLETION not in line.dump_token_type_ids():
```
Comment (snova-virens) [inline generative_data_prep/processors/article_tokenizer.py:198]: move the sum here.

File: generative_data_prep/processors/metrics.py [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+
+    @property
+    def average_completion_length(self) -> float:
+        """The average number of tokens per completion."""
+        return self.completion_tokens / self.articles
+
+    @property
+    def data_utilization(self) -> float:
+        """What percent of the input dataset tokens are actually included."""
+        return (self.prompt_tokens + self.completion_tokens) / self.input_tokens
```
Comment (snova-virens) [inline generative_data_prep/processors/metrics.py:85]: shouldn't this just be self.output_tokens / self.input_tokens?

Comment (snova-virens) [reply generative_data_prep/processors/metrics.py:85]: ohhh u're excluding padding, that's smart

File: generative_data_prep/processors/metrics.py [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+
+This module implements ArticleTokenizer.
+
+Article tokenizer takes any input text or jsonl, runs tokenization using
```
Comment (snova-virens) [inline generative_data_prep/processors/metrics.py:18]: Please fix this

**Review by snova-virens — APPROVED** [review]
Looks great!

**Review by snova-virens — APPROVED** [review]
Thanks Zoli!

**Review by snova-ranjanl — APPROVED** [review]

**Pull request merged**